src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -1897,10 +1897,14 @@ static Node* split_flow_path(PhaseGVN *phase, PhiNode *phi) {
 
 // Returns the BasicType of a given convert node and a type, with special handling to ensure that conversions to
 // and from half float will return the SHORT basic type, as that wouldn't be returned typically from TypeInt.
-static BasicType get_convert_type(Node* convert, const Type* type) {
+static BasicType get_convert_type(Node* convert, const Type* type, const bool is_source=false) {
   int convert_op = convert->Opcode();
-  if (type->isa_int() && (convert_op == Op_ConvHF2F || convert_op == Op_ConvF2HF)) {
-    return T_SHORT;
+  if (type->isa_int()) {
+    if (is_source && convert_op == Op_ConvHF2I) {
+      return T_SHORT;
+    } else if (convert_op == Op_ConvHF2F || convert_op == Op_ConvF2HF || convert_op == Op_ConvHF2D || convert_op == Op_ConvHF2L) {
+      return T_SHORT;
+    }
```

Comment:
This transformation is moving conversion across the Phi node thereby saving expensive conversion instructions.

`Phi (ConvHF2F src1) (ConvHF2F src2) =>   ConvHF2F  (Phi src1  src2)
`

For the Op_ConvF2HF case, type parameter will be of float type and thus we will never enter control flow between L1905 - L1907. 

I think it will be harmless if we remove convert_op == Op_ConvF2HF check...

Reply:
We can also update existing test for new conversion

test/hotspot/jtreg/compiler/c2/irTests/TestPhiDuplicatedConversion.java

Reply:
> This transformation is moving conversion across the Phi node thereby saving expensive conversion instructions.
> 
> `Phi (ConvHF2F src1) (ConvHF2F src2) => ConvHF2F (Phi src1 src2) `
> 
> For the Op_ConvF2HF case, type parameter will be of float type and thus we will never enter control flow between L1905 - L1907.
> 
> I think it will be harmless if we remove convert_op == Op_ConvF2HF check...

Hi @jatin-bhateja , I think that check should be there. The lines 1905-1907 will be reachable when get_convert_type() is called again with the dest_type (`short`) when it is supposed to return `short`. If this check is not there, it will return `T_INT`. For the following example - 
`sout[i] = rng.nextBoolean() ? Float.floatToFloat16(a) : Float.floatToFloat16(b)
`
it will generate `ConvF2I` instead of `ConvF2HF`.

Reply:
I have added tests in `TestPhiDuplicatedConversion.java`